### DISCUSSION TITLE
PnP Search extensibility - Custom Web component with user query

### DISCUSSION THREAD

**Discussion by mvgadagi** [microsoft-search/pnp-modern-search#3086] [Q&A] [answered]
Hello,

Currently, I am looking for a simple example to demonstrate the below use case.
I want to design a custom web component to show some results from other data sources than Microsoft Graph & Sharepoint.
So I am thinking of developing a custom web component with making a call to external API from the user query. Could you please assist me?

1) How to capture the user query inside the custom web component, I tried this below to see if there is a user query inside any of these contexts but I couldn't find any.

```
props.context.serviceScope = this._serviceScope;
props.context.pageContext = this._serviceScope.consume(PageContext.serviceKey);
```

2) Please suggest me if you have any other suggestions on this approach ( I already saw a document about Custom Data source, but in my case, I just need to bring one user card based on the user query, so I am thinking this approach now)

Thanks a lot in advance !!

**Comment by patrikhellgren** [comment, marked as answer]
You can pass the input query text to your custom web component from the handlebars layout template like below. If you want to pass other stuff from the search context, have a look at the debug layout template and find what you want there and use it in the same way as ```inputQueryText```.

### Custom React Component
``` TypeScript
export interface ICustomReactComponentProps {
  inputQueryText: string;
}

export function CustomReactComponent(props: ICustomReactComponentProps ): React.ReactElement<ICustomReactComponentProps > {
    return <div>
            {props.inputQueryText}
           </div>
}
```

### Custom Web Component
``` TypeScript
  public async connectedCallback() {
    let props = this.resolveAttributes();
    const customReactComponent: React.ReactElement<ICustomReactComponentProps > = React.createElement(
      CustomReactComponent,
      {
        ...props
      }
    );

    ReactDOM.render(customReactComponent, this);
  }    
```

### Custom Handlebars Layout
``` HTML
<custom-react-componment
    data-input-query-text="{{inputQueryText}}"
></custom-react-componment>
```

**Comment by mvgadagi** [comment]
Hello @patrikhellgren Thanks a lot, It worked now, and also I understood the flow of your answer.
Thanks again !!
